src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -429,3 +433,252 @@ deriveFunctor mn tyConNm = do
+
+          foldMapArg :: (Bool, Expr) -> Expr
+          foldMapArg (reuseFoldr, next)
+            | reuseFoldr = App (App foldMapVar fVar) next
+            | otherwise = App fVar next
+
+          appendArgs :: Expr -> Expr -> Expr
+          appendArgs l r = App (App appendVar l) r
+
+        case nonEmpty ctorArgs of
+          Nothing ->
+            memptyVar
+          Just args ->
+            foldl1 appendArgs $ fmap foldMapArg args
+
+    mkCtorClauseWith
+      :: Text
+      -> ([(Bool, Expr)] -> Expr)
+      -> (ProperName 'ConstructorName, [SourceType])
+      -> m CaseAlternative
+    mkCtorClauseWith iTyName buildExpr (ctorName, ctorTys) = do
+      idents <- replicateM (length ctorTys) (freshIdent "v")
+      ctorTys' <- mapM replaceAllTypeSynonyms ctorTys
+      let
+        args = catMaybes $ zipWith (indicateRelevantArgs iTyName) idents ctorTys'
+        finalExpr = buildExpr args
+        caseBinder = mkCtorBinder mn ctorName $ map mkBinder idents
+      return $ CaseAlternative [caseBinder] (unguarded finalExpr)
+
+    -- | Nothing = irrelevant
+    -- | Just (True, _) = recursive call to @fold@ function
+    -- | Just (False, _) = call @f@ function on arg
+    indicateRelevantArgs :: Text -> Ident -> SourceType -> Maybe (Bool, Expr)
+    indicateRelevantArgs iTyName ident = goType
+      where
+      -- Note: this function is based on the Functor `goType` above.
+      -- I'm not sure whether `ConstrainedType` should be included.
+      goType :: SourceType -> Maybe (Bool, Expr)
+      goType = \case
+        -- argument matches the index type
+        TypeVar _ t | t == iTyName -> pure (False, mkVar ident)
+
+        -- quantifiers
+        (ForAll _ scopedVar _ t _) | scopedVar /= iTyName -> goType t
+
+        -- constraints
+        (ConstrainedType _ _ t) -> goType t
+
+        -- under a `* -> *`, just assume Foldable for now
+        (TypeApp _ _ t) -> do
+          (\(_, expr) -> (True, expr)) <$> goType t
+
+        -- otherwise invalid arg
+        _ -> Nothing
+
+deriveTraversable
+  :: forall m
+   . MonadError MultipleErrors m
+  => MonadState CheckState m
+  => MonadSupply m
+  => ModuleName
+  -> ProperName 'TypeName
+  -> m [(PSString, Expr)]
+deriveTraversable mn tyConNm = do
+  (_, _, tys, ctors) <- lookupTypeDecl mn tyConNm
+  case reverse tys of
+    [] ->
+      throwError . errorMessage $ KindsDoNotUnify (kindType -:> kindType) kindType
+    ((iTy, _) : _) -> do
+      traverseFn <- mkTraverseFn iTy ctors
+      sequenceFn <- mkSequenceFn
+      pure [(Traversable.traverse, traverseFn), (Traversable.sequence, sequenceFn)]
+  where
+    traverseVar = mkRef Traversable.identTraverse
+    identityVar = mkRef Prelude.identIdentity
+    mapVar = mkRef Prelude.identMap
+    applyVar = mkRef Prelude.identApply
+    pureVar = mkRef Prelude.identPure
+
+    mkSequenceFn
+      :: m Expr
+    mkSequenceFn = do
+      t <- freshIdent "t"
+      pure $ lam t $ App (App traverseVar identityVar) (mkVar t)
```

Comment:
I'm wondering if this should be eta-reduced.

Reply:
Good question. I would say no, given that there are some known issues with eta-reduced definitions with e.g. recursive typeclass definitions.